dataset/shape.cpp

```diff
@@ -141,4 +141,232 @@ Dataset resize(const DatasetConstView &d, const Dim dim,
   return result;
 }
 
+namespace {
+
+/// Stack dims for reshaping one dim into multiple dims
+///
+/// Go through the old dims and:
+/// - if the dim does not equal the dim that is being stacked, copy dim/shape
+/// - if the dim equals the dim to be stacked, replace by stack of new dims
+const Dimensions stack_dims(const Dimensions &old_dims, const Dim from_dim,
+                            const Dimensions &to_dims) {
+  Dimensions new_dims;
+  for (const auto dim : old_dims.labels())
+    if (dim != from_dim)
+      new_dims.addInner(dim, old_dims[dim]);
+    else
+      for (const auto lab : to_dims.labels())
+        new_dims.addInner(lab, to_dims[lab]);
+  return new_dims;
+}
+
+/// Unstack/flatten dims for reshaping multiple dims into one
+///
+/// Go through the old dims and:
+/// - if the dim is contained in the list of dims to be flattened, add the new
+///   dim once
+/// - if not, copy the dim/shape
+const Dimensions unstack_dims(const Dimensions &old_dims,
+                              const Dimensions &from_dims, const Dim to_dim) {
+  Dimensions new_dims;
+  for (const auto dim : old_dims.labels())
+    if (from_dims.contains(dim)) {
+      if (!new_dims.contains(to_dim))
+        new_dims.addInner(to_dim, from_dims.volume());
```

Comment:
What if we are, e.g., unstacking a mask that does not depend on all the stacked dims? Isn't `from_dims.volume()` too much in this case? Shouldn't it only consider dimensions (and the subvolume) actually contained in`old_dims`?

Reply:
hmm i'm not exactly sure what you mean by the subvolume contained in `old_dims`. It's because `old_dims` doesn't have the full volume that you need to broadcast to the `from_dims`.

In this picture, we flatten dimensions `y` and `z` into a dim `a`. There are 1d masks along the `x`, `y`, and `z` dimensions.
Masks `y` and `z` have to be broadcast to dims `['y', 'z']` in order to be flattened. They are not being broadcast to the full `['x', 'y', 'z']` since `from_dims` is `['y', 'z']`.
Mask `x` is unchanged.

Maybe I am missing something?
![masks](https://user-images.githubusercontent.com/39047984/110159194-ae98c980-7dea-11eb-9468-c730c60bf54a.png)

![Screenshot at 2021-03-05 19-47-59](https://user-images.githubusercontent.com/39047984/110160009-ba38c000-7deb-11eb-8359-313e4fcbf671.png)